src/mo_frontend/typing.ml

```diff
@@ -1738,7 +1738,10 @@ and infer_exp'' env exp : T.typ =
     let t0 = check_AwaitCap env "await" exp.at in
     let t1 = infer_exp_promote env exp1 in
     (try
-       let (t2, t3) = T.as_async_sub s t0 t1 in
+       let s1 = match s, t1 with
+         | T.(Cmp, Async(Fut, _, _)) -> T.Fut (* we can await* an async *)
+         | _ -> s in
+       let (t2, t3) = T.as_async_sub s1 t0 t1 in
```

Comment:
Maybe this would be cleaner if T.as_async_sub returned the sort of t1  and then you act accordingly:

let (s1,t2,t3) = T.as_async _sub s t0 t1 in
...
if (s = T.Fut && s1 == T.Cmp)  then error....;

Reply:
My GADT suggestion would eliminate that:
``` Haskell
data kind Mode = Intro | Elim
data AsyncSort Mode where
    Fut :: AsyncSort both
    FutNoBlock :: AsyncSort Elim
    Cmp :: AsyncSort both
```
and then `AsyncE (AsyncSort Intro, ...)` and `AwaitPrim (AsyncSortElim, ...)`